### PR TITLE
nrf_modem_lib: `_ON_INIT` and `_ON_SHUTDOWN` macros

### PR DESCRIPTION
**Pull request by lemrey**
Add a compile-time callbacks for 
- modem initialization status: `NRF_MODEM_LIB_ON_INIT` and `NRF_MODEM_LIB_ON_SHUTDOWN`
- modem functional mode changes: `LTE_LC_ON_CFUN`

Add sample to showcase both.

This achieves a few things; first, that any steps that a library or the application needs to perform whenever the modem is initialized/shutdown or in a certain functional mode, are performed (and repeated) automatically whenever necessary (e.g. MAGPIO configuration, AT notification subscriptions). Also, we loose the coupling between the different parts of the application and ultimately won't have to rely on a chain of `_ON_SYS_INIT` Kconfig and `#ifdef` to execute these setup steps.

Other than introducing the macros, there a couple of changes to `lte_lc`.
- run configuration steps in the modem library initialization callback [b9e9705](https://github.com/nrfconnect/sdk-nrf/pull/7283/commits/b9e970525bd81dd5fb7dd3d0f424490a43191c5b)
- run `+CFUN=0` in modem library shutdown callback [129ef5b](https://github.com/nrfconnect/sdk-nrf/pull/7283/commits/129ef5be5420be78b4da7242360eec134f75290c)

Ref old PR: #6584

### CODE REVIEW

File: samples/nrf9160/modem_callbacks/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/CMakeLists.txt:2]:
```suggestion
# Copyright (c) 2022 Nordic Semiconductor
```

File: samples/nrf9160/modem_callbacks/README.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
+.. _modem_callbacks_sample:
+
+nRF9160: Modem callbacks
+########################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The Modem callbacks sample demonstrates how to setup callbacks for Modem library initialization and shutdown calls, using the :ref:`nrf_modem_lib_readme` and how to setup a callback on changes to the modem function mode using the :ref:`lte_lc_readme` library.
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:10]:
```suggestion
The Modem callbacks sample demonstrates how to setup callbacks for Modem library initialization and shutdown calls, using the :ref:`nrf_modem_lib_readme` and how to setup a callback on changes to the modem functional mode using the :ref:`lte_lc_readme` library.
```

File: samples/nrf9160/modem_callbacks/README.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+Testing
+=======
+
+|test_sample|
+
+1. |connect_kit|
+#. |connect_terminal|
+#. Power on or reset your nRF9160 DK.
+#. Observe that the sample starts, creates and configures a PDP context, and then activates a PDN connection.
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:47]: Is this PDP/PDN stuff leftovers from a different sample?

File: samples/nrf9160/modem_callbacks/README.rst [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+Sample output
+=============
+
+The sample shows the following output, which may vary based on the network provider:
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:52]: Probably the output shouldn't vary based on the network provider?

File: samples/nrf9160/modem_callbacks/prj.conf [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/prj.conf:2]:
```suggestion
# Copyright (c) 2022 Nordic Semiconductor ASA
```

File: samples/nrf9160/modem_callbacks/src/main.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		return;
+	}
+
+	printk("Changing function mode\n");
+
+	err = lte_lc_init_and_connect();
+	if (err) {
+		printk("lte_lc_init_and_connect() failed, err %d\n", err);
+	}
+
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/src/main.c:67]: Extra empty line.

File: samples/nrf9160/modem_callbacks/src/main.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+static void on_modem_shutdown(void *ctx)
+{
+	printk("> Shutting down\n");
+}
+
+LTE_LC_ON_CFUN(cfun_monitor, on_cfun, NULL);
+
+static void on_cfun(enum lte_lc_func_mode mode, void *ctx)
+{
+	printk("> Function mode has changed to %d\n", mode);
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/src/main.c:30]:
```suggestion
	printk("> Functional mode has changed to %d\n", mode);
```

File: samples/nrf9160/modem_callbacks/README.rst [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+  Modem callbacks sample started
+  Initializing modem library
+  > Initialized with value 0
+  Shutting down modem library
+  > Shutting down
+  Initializing modem library (again)
+  > Initialized with value 0
+  Changing function mode
+  > Function mode has changed to 1
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:64]:
```suggestion
  Changing functional mode
  > Functional mode has changed to 1
```

File: samples/nrf9160/modem_callbacks/README.rst [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  Shutting down modem library (again)
+  > Function mode has changed to 0
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:66]:
```suggestion
  > Functional mode has changed to 0
```

File: samples/nrf9160/modem_callbacks/src/main.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+	printk("Initializing modem library (again)\n");
+
+	err = nrf_modem_lib_init(NORMAL_MODE);
+	if (err) {
+		printk("libmodem initialization failed, err %d\n", err);
+		return;
+	}
+
+	printk("Changing function mode\n");
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/src/main.c:61]:
```suggestion
	printk("Changing functional mode\n");
```

File: samples/nrf9160/modem_callbacks/src/main.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	err = nrf_modem_lib_init(NORMAL_MODE);
+	if (err) {
+		printk("Modem initialization failed, err %d\n", err);
+		return;
+	}
+
+	printk("Shutting down modem library\n");
+	err = nrf_modem_lib_shutdown();
+	if (err) {
+		printk("Modem initialization failed, err %d\n", err);
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/src/main.c:49]:
```suggestion
		printk("Shutting down modem failed, err %d\n", err);
```

File: samples/nrf9160/modem_callbacks/src/main.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	if (err) {
+		printk("lte_lc_init_and_connect() failed, err %d\n", err);
+	}
+
+
+	printk("Shutting down modem library (again)\n");
+
+	err = nrf_modem_lib_shutdown();
+	if (err) {
+		printk("libmodem initialization failed, err %d\n", err);
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/src/main.c:73]:
```suggestion
		printk("Shutting down modem failed, err %d\n", err);
```

File: doc/nrf/libraries/modem/lte_lc.rst [hunk 13]
```diff
@@ -230,6 +230,33 @@ You can use the following configuration options to configure the connection fall
...
+
+.. code-block:: C
+
+  /* define callback */
+  LTE_LC_ON_CFUN(cfun_hook, on_cfun, NULL);
+
+  /* callback implementation */
+  static void on_cfun(enum lte_lc_func_mode mode, void *context)
+  {
+      printk("Function mode changed to %d\n", mode);
```
Comment (tokangas) [inline doc/nrf/libraries/modem/lte_lc.rst:251]:
```suggestion
      printk("Functional mode changed to %d\n", mode);
```

**Review by tokangas — COMMENTED** [review]
LGTM, only some cosmetic changes.

I was wondering do we need a new sample just for this, because this functionality should be taken into use in several existing samples?

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 14]
```diff
@@ -23,6 +23,10 @@ The library wrapper eases the task of initializing the Modem library by automati
 :ref:`partition_manager` is the component that reserves the RAM memory for the shared memory regions used by the Modem library.
 For more information, see :ref:`partition_mgr_integration`.
 
+The library wrapper also provides callbacks for the initialization and shutdown operations.
+The application can setup a callback for :c:func:`nrf_modem_lib_init` calls using the :c:macro:`NRF_MODEM_LIB_ON_INIT` macro, and a callback for :c:func:`nrf_modem_lib_shutdown` calls using the :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` macro.
+These compile-time callbacks allow any part of the application to perform any setup steps that require the modem to be a certain state. Futher, they ensure that said steps will be repeated any time another part of the application turns the modem on or off.
```
Comment (eivindj-nordic) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:28]: Not sure how relevant it is, but should this specify that the init callbacks are called after the modem is initialized and the shutdown callbacks are called before the modem is shut down?

**Comment by lemrey** [comment]
I though it made sense to have a sample to show how to use this functionality in the simplest way possible.

File: samples/nrf9160/modem_callbacks/README.rst [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+The sample shows the following output, which may vary based on the network provider:
+
+.. code-block:: console
+
+  Modem callbacks sample started
+  Initializing modem library
+  > Initialized with value 0
+  Shutting down modem library
+  > Shutting down
+  Initializing modem library (again)
```
Comment (eivindj-nordic) [inline samples/nrf9160/modem_callbacks/README.rst:61]: This (and the next lines) belongs to the [lib: lte_lc: add LTE_LC_ON_CFUN macro](https://github.com/nrfconnect/sdk-nrf/pull/7283/commits/27b48f723b65d7f7067c0bcb3ac672348f5d05e8) commit.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 14 above)
Comment (lemrey) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:28]: Good point, I'll mention it here and in the doxygen.

File: samples/nrf9160/modem_callbacks/sample.yaml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+sample:
+  name: modem init shutdown
+tests:
+  samples.nrf9160.pdn:
```
Comment (rlubos) [inline samples/nrf9160/modem_callbacks/sample.yaml:4]: This should be updated.

**Review by rlubos — COMMENTED** [review]
Nice

File: include/modem/nrf_modem_lib.h [hunk 17]
```diff
@@ -113,7 +113,7 @@ struct nrf_modem_lib_shutdown_cb {
  * When nrf_modem_lib_shutdown() is called a thread can call this function to be
  * woken up next time nrf_modem_lib_init() is called.
  */
-void nrf_modem_lib_shutdown_wait(void);
+__deprecated void nrf_modem_lib_shutdown_wait(void);
```
Comment (jtguggedal) [inline include/modem/nrf_modem_lib.h:116]: I might be missing something here, but looks like the changelog comment on `nrf_modem_lib_get_init_ret` doesn't match the function marked as deprecated, `nrf_modem_lib_shutdown_wait`

File: samples/nrf9160/modem_callbacks/README.rst [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+Testing
+=======
+
+|test_sample|
+
+1. |connect_kit|
+#. |connect_terminal|
+#. Power on or reset your nRF9160 DK.
+#. Observe that the sample starts, initializes and shuts down the modem.
+#. Observe that the sample initializes the modem, changes functional modem and shuts down the modem.
```
Comment (tokangas) [inline samples/nrf9160/modem_callbacks/README.rst:48]:
```suggestion
#. Observe that the sample initializes the modem, changes functional mode and shuts down the modem.
```

File: include/modem/nrf_modem_lib.h (hunk 17 above)
Comment (lemrey) [reply include/modem/nrf_modem_lib.h:116]: I plan to deprecate that function too btw, but I'll will be a separate PR.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 19]
```diff
@@ -318,6 +319,22 @@ Modem libraries
...
+
+  * :ref:`lte_lc_readme` library:
+
+    * Added :c:macro:`LTE_LC_ON_CFUN` macro for compile-time registration of callbacks on modem functional mode changes using :c:func:`lte_lc_func_mode_set`.
+
+* Deprecated:
+
+  * :ref:`nrf_modem_lib_readme` library:
+
+    * :c:func:`nrf_modem_lib_shutdown_wait`, in favor of :c:macro:`NRF_MODEM_LIB_ON_INIT`.
```
Comment (divipillai) [inline doc/nrf/releases/release-notes-changelog.rst:337]:
```suggestion
  * :ref:`nrf_modem_lib_readme` library:
    
    * Added :c:macro:`NRF_MODEM_LIB_ON_INIT` macro for compile-time registration of callbacks on modem initialization.
    * Added :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` macro for compile-time registration of callbacks on modem de-initialization.
    * Removed :c:func:`nrf_modem_lib_shutdown_wait` function, in favor of :c:macro:`NRF_MODEM_LIB_ON_INIT`.
    
  * :ref:`lte_lc_readme` library:

    * Added :c:macro:`LTE_LC_ON_CFUN` macro for compile-time registration of callbacks on modem functional mode changes using :c:func:`lte_lc_func_mode_set`.
```

File: samples/nrf9160/modem_callbacks/README.rst [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+.. include:: /includes/spm.txt
+
+Overview
+********
+
+The sample registers three callbacks at compile time: :c:func:`on_modem_init`, :c:func:`on_modem_shutdown` and :c:func:`on_cfun` using the :c:macro:`NRF_MODEM_LIB_ON_INIT`, :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` and :c:macro:`LTE_LC_ON_CFUN` macros respectively.
+Next, the sample initializes and shuts down the :ref:`nrfxlib:nrf_modem`, triggering :c:func:`on_modem_init` and :c:func:`on_modem_shutdown`.
+Then, the sample initializes the modem library again, changes functional mode using the :c:func:`lte_lc_func_mode_set` function in the :ref:`lte_lc_readme` library, and shuts down the modem library again.
+This triggers the :c:func:`on_modem_init`, :c:func:`on_cfun` and :c:func:`on_modem_shutdown` callbacks.
```
Comment (divipillai) [inline samples/nrf9160/modem_callbacks/README.rst:29]:
```suggestion
The sample performs the following operations:

1. Registers callbacks during compile time for :c:func:`on_modem_init`, :c:func:`on_modem_shutdown` and :c:func:`on_cfun` functions using the :c:macro:`NRF_MODEM_LIB_ON_INIT`, :c:macro:`NRF_MODEM_LIB_ON_SHUTDOWN` and :c:macro:`LTE_LC_ON_CFUN` macros respectively.
#. Initializes and shuts down the :ref:`nrfxlib:nrf_modem` by triggering :c:func:`on_modem_init` and :c:func:`on_modem_shutdown` functions.
#. Initializes the modem library.
#. Changes functional mode using the :c:func:`lte_lc_func_mode_set` function in the :ref:`lte_lc_readme` library
#. Shuts down the modem library.

This triggers the callbacks for :c:func:`on_modem_init`, :c:func:`on_cfun` and :c:func:`on_modem_shutdown` functions.
```

**Review by simensrostad — APPROVED** [review]
Looks good to me! 🚗

**Review by tokangas — APPROVED** [review]
Same here!

**Review by eivindj-nordic — APPROVED** [review]

**Review by divipillai — APPROVED** [review]

**Pull request merged**